File: desktop/split_tunnel_injector.go
package main

import (
	"fmt"
	"log"
	"strings"
)

// SplitTunnelInject patches a config blob with split-tunnel bypass
// directives derived from the pool's PoolSplitTunnel settings.
// Mirror of Android's data.SplitTunnelInjector.
//
//   wireguard: replace [Peer] AllowedIPs with the complement.
//   openvpn:   prepend route + route-ipv6 net_gateway directives
//              plus pull-filter ignore for matching server pushes.
//   ipsec:     no-op + warning. Traffic selectors require server
//              cooperation.
//
// Returns (patchedConfig, applied, skipReason).
//   applied=true  -> caller can log success; patchedConfig is the
//                    modified blob.
//   applied=false -> skipReason explains why (member config not
//                    full-tunnel, IPSec, etc); patchedConfig equals
//                    the input.
func SplitTunnelInject(configContent string, protocol string, st PoolSplitTunnel) (string, bool, string) {
	if !st.IsActive() {
		return configContent, false, ""
	}

	parsed := make([]Cidr, 0, len(st.BypassCidrs))
	for _, raw := range st.BypassCidrs {
		c, err := ParseCidr(raw)
		if err == nil {
			parsed = append(parsed, c)
		}
	}
	if st.ExcludePrivateNetworks {
		parsed = append(parsed, PrivateNetworks...)
	}
	if len(parsed) == 0 {
		return configContent, false, "no valid CIDRs after parse"
	}

	switch protocol {
	case "wireguard":
		return patchWGSplit(configContent, parsed)
	case "openvpn":
		return patchOVPNSplit(configContent, parsed)
	case "ipsec":
		log.Printf("split tunnel skipped: IPSec doesn't support client-side traffic-selector narrowing")
		return configContent, false, "IPSec member - client-side split tunnel not supported"
	default:
		return configContent, false, "unknown protocol"
	}
}

// patchWGSplit: WireGuard config-text patch. Find AllowedIPs lines
// in any [Peer] section, parse, compute complement, replace.
// Disable if no full-tunnel coverage in original.
func patchWGSplit(content string, bypass []Cidr) (string, bool, string) {
	lines := strings.Split(content, "\n")
	allowedIdxs := []int{}
	for i, l := range lines {
		t := strings.TrimSpace(l)
		if !strings.HasPrefix(strings.ToLower(t), "allowedips") {
			continue
		}
		if !strings.Contains(t, "=") {
			continue
		}
		allowedIdxs = append(allowedIdxs, i)
	}
	if len(allowedIdxs) == 0 {
		return content, false, "no AllowedIPs line in config"
	}

	// Aggregate existing AllowedIPs CIDRs across the lines.
	existing := []Cidr{}
	for _, idx := range allowedIdxs {
		rhs := strings.TrimSpace(lines[idx][strings.IndexByte(lines[idx], '=')+1:])
		for _, raw := range strings.Split(rhs, ",") {
			c, err := ParseCidr(strings.TrimSpace(raw))
			if err != nil {
				continue
			}
			existing = append(existing, c)
		}
	}

	hasV4Univ := false
	hasV6Univ := false
	for _, c := range existing {
		if c.Prefix != 0 {
			continue
		}
		if c.IsV4() {
			hasV4Univ = true
		} else {
			hasV6Univ = true
		}
	}

	if !hasV4Univ && !hasV6Univ {
		log.Printf("split tunnel skipped: AllowedIPs is not full-tunnel " +
			"(no 0.0.0.0/0 or ::/0) - existing routes would conflict")
		return content, false, "member config is not full-tunnel; pool-level bypass cannot apply safely"
	}

	// Compute complement only for the families present.
	effectiveBypass := make([]Cidr, 0, len(bypass))
	for _, c := range bypass {
		if (c.IsV4() && hasV4Univ) || (!c.IsV4() && hasV6Univ) {
			effectiveBypass = append(effectiveBypass, c)
		}
	}
	complement := SubtractFromUniverse(effectiveBypass)
	filtered := complement[:0]
	for _, c := range complement {
		if (c.IsV4() && hasV4Univ) || (!c.IsV4() && hasV6Univ) {
			filtered = append(filtered, c)
		}
	}
	if len(filtered) == 0 {
		log.Printf("split tunnel skipped: bypass covers full universe")
		return content, false, "bypass set covers entire address space"
	}

	parts := make([]string, 0, len(filtered))
	for _, c := range filtered {
		parts = append(parts, c.String())
	}
	replacement := "AllowedIPs = " + strings.Join(parts, ", ")

	// Replace first AllowedIPs line; remove any subsequent ones
	// (multi-line entries collapsed into one for readability).
	out := make([]string, 0, len(lines))
	firstReplaced := false
	for i, l := range lines {
		isAllowed := false
		for _, idx := range allowedIdxs {
			if idx == i {
				isAllowed = true
				break
			}
		}
		if !isAllowed {
			out = append(out, l)
			continue
		}
		if !firstReplaced {
			out = append(out, replacement)
			firstReplaced = true
		}
		// subsequent AllowedIPs lines: drop
	}
	log.Printf("WG split tunnel applied: %d bypass CIDRs, %d resulting AllowedIPs entries",
		len(effectiveBypass), len(filtered))
	return strings.Join(out, "\n"), true, ""
}

// patchOVPNSplit: prepend route net_gateway directives + pull-filter
// ignore for matching server pushes.
func patchOVPNSplit(content string, bypass []Cidr) (string, bool, string) {
	var sb strings.Builder
	sb.WriteString("# Privycs split tunnel: client-side bypass routes\n")
	for _, c := range bypass {
		ip := strings.SplitN(c.String(), "/", 2)[0]
		if c.IsV4() {
			fmt.Fprintf(&sb, "pull-filter ignore \"route %s\"\n", ip)
		} else {
			fmt.Fprintf(&sb, "pull-filter ignore \"route-ipv6 %s\"\n", ip)
		}
	}
	for _, c := range bypass {
		if c.IsV4() {
			ip := strings.SplitN(c.String(), "/", 2)[0]
			mask := prefixToIPv4Mask(c.Prefix)
			fmt.Fprintf(&sb, "route %s %s net_gateway\n", ip, mask)
		} else {
			fmt.Fprintf(&sb, "route-ipv6 %s net_gateway_ipv6\n", c.String())
		}
	}
	sb.WriteString(content)
	log.Printf("OpenVPN split tunnel applied: %d bypass CIDRs", len(bypass))
	return sb.String(), true, ""
}

func prefixToIPv4Mask(prefix int) string {
	if prefix < 0 || prefix > 32 {
		return "255.255.255.255"
	}
	var mask uint32
	if prefix > 0 {
		mask = 0xFFFFFFFF << (32 - prefix)
	}
	return fmt.Sprintf("%d.%d.%d.%d",
		(mask>>24)&0xFF, (mask>>16)&0xFF, (mask>>8)&0xFF, mask&0xFF)
}

File: android/app/src/main/java/com/privycs/vpn/data/SplitTunnelInjector.kt
package com.privycs.vpn.data

import android.util.Log
import com.privycs.vpn.data.models.PoolSplitTunnel
import com.privycs.vpn.data.models.VpnProtocol

/**
 * Patches a pool member's config blob with split-tunnel bypass
 * directives derived from the pool's [PoolSplitTunnel] settings.
 *
 * Per-protocol semantics:
 *
 *   WireGuard:  rewrite the [Peer] section's `AllowedIPs = ...`
 *               line as the complement of the bypass set against
 *               the original AllowedIPs. If the original is not
 *               full-tunnel (0.0.0.0/0 OR ::/0), the bypass is
 *               disabled with a log warning - per the user's
 *               decision in v0.9.11.55 design discussion.
 *
 *   OpenVPN:    append `route X mask Y net_gateway` (IPv4) and
 *               `route-ipv6 X::/n net_gateway_ipv6` (IPv6) per
 *               bypass CIDR. Plus `pull-filter ignore "route X"`
 *               directives for any matching server-pushed routes
 *               that would otherwise compete.
 *
 *   IPSec:      no-op + warning. IPSec traffic selectors are
 *               negotiated with the server; client-side narrowing
 *               isn't reliable across providers.
 *
 * Empty / disabled config returns the input unchanged.
 */
object SplitTunnelInjector {

    private const val TAG = "SplitTunnel"

    /**
     * Result type so callers can surface a "split tunnel disabled
     * because the member config isn't full-tunnel" message in the
     * UI / logs without needing to re-parse the config.
     */
    data class InjectResult(
        val patched: String,
        val applied: Boolean,
        val skippedReason: String? = null
    )

    /**
     * Apply the split-tunnel config to the given member config text.
     * No-op when the split-tunnel is disabled (toggle off + no
     * bypass CIDRs).
     */
    fun inject(
        configContent: String,
        protocol: VpnProtocol,
        splitTunnel: PoolSplitTunnel
    ): InjectResult {
        if (!splitTunnel.isActive()) {
            return InjectResult(configContent, applied = false)
        }
        // Build the effective bypass CIDR set: user-typed entries
        // (skipping malformed strings - the UI validates but we
        // double-check here) + RFC1918/IPv6-ULA if enabled.
        val parsed = mutableListOf<CidrMath.Cidr>()
        for (s in splitTunnel.bypassCidrs) {
            CidrMath.parse(s)?.let { parsed.add(it) }
        }
        if (splitTunnel.excludePrivateNetworks) {
            parsed.addAll(CidrMath.PRIVATE_NETWORKS)
        }
        if (parsed.isEmpty()) {
            // Nothing valid to apply. Treat as disabled so we don't
            // accidentally rewrite AllowedIPs to 0.0.0.0/0+::/0
            // (no-change) and trigger a needless tunnel rebuild.
            return InjectResult(configContent, applied = false,
                skippedReason = "no valid CIDRs after parse")
        }

        return when (protocol) {
            // AmneziaWG shares the WG .conf grammar — same AllowedIPs
            // narrowing applies unchanged.
            VpnProtocol.WIREGUARD, VpnProtocol.AMNEZIAWG -> patchWireGuard(configContent, parsed)
            VpnProtocol.OPENVPN -> patchOpenVpn(configContent, parsed)
            VpnProtocol.IPSEC -> {
                Log.w(TAG, "split tunnel skipped: IPSec doesn't support client-side traffic-selector narrowing")
                InjectResult(configContent, applied = false,
                    skippedReason = "IPSec member - client-side split tunnel not supported")
            }
        }
    }

    /**
     * WireGuard patch: parse AllowedIPs from [Peer], compute
     * complement, replace. Disable if AllowedIPs doesn't include
     * a full-tunnel route.
     */
    private fun patchWireGuard(
        content: String,
        bypass: List<CidrMath.Cidr>
    ): InjectResult {
        val lines = content.lines().toMutableList()

        // Find AllowedIPs lines (peer section may have one or
        // multiple AllowedIPs lines, each comma-separated).
        val allowedIPLineIdxs = lines.indices.filter {
            lines[it].trim().startsWith("AllowedIPs", ignoreCase = true) &&
                    lines[it].contains('=')
        }
        if (allowedIPLineIdxs.isEmpty()) {
            return InjectResult(content, applied = false,
                skippedReason = "no AllowedIPs line in config")
        }

        // Aggregate all existing AllowedIPs CIDRs across the lines.
        val existingCidrs = mutableListOf<CidrMath.Cidr>()
        for (idx in allowedIPLineIdxs) {
            val rhs = lines[idx].substringAfter('=').trim()
            for (raw in rhs.split(',')) {
                val c = CidrMath.parse(raw.trim()) ?: continue
                existingCidrs.add(c)
            }
        }

        // v4 universe present? (0.0.0.0/0)
        val hasV4Universe = existingCidrs.any { it.isV4 && it.prefix == 0 }
        // v6 universe present? (::/0)
        val hasV6Universe = existingCidrs.any { !it.isV4 && it.prefix == 0 }

        if (!hasV4Universe && !hasV6Universe) {
            // Member config is already a custom split-tunnel
            // (provider-specific routes). User decision: disable
            // bypass with warning rather than risk unpredictable
            // route layering.
            Log.w(TAG, "split tunnel skipped: AllowedIPs is not full-tunnel " +
                    "(no 0.0.0.0/0 or ::/0) - existing routes would conflict")
            return InjectResult(content, applied = false,
                skippedReason = "member config is not full-tunnel; pool-level bypass cannot apply safely")
        }

        // Compute complement only for the families present in the
        // existing AllowedIPs. If existing is v4-only, leave v6
        // out of the result (we don't want to add ::/0 silently).
        val effectiveBypass = bypass.filter {
            (it.isV4 && hasV4Universe) || (!it.isV4 && hasV6Universe)
        }
        val complement = CidrMath.subtractFromUniverse(effectiveBypass)
            .filter {
                (it.isV4 && hasV4Universe) || (!it.isV4 && hasV6Universe)
            }

        if (complement.isEmpty()) {
            // Pathological: bypass covers the entire universe(s).
            // Skip rather than emit "AllowedIPs = " which the WG
            // parser may reject or treat as "block all".
            Log.w(TAG, "split tunnel skipped: bypass covers full universe, would result in empty AllowedIPs")
            return InjectResult(content, applied = false,
                skippedReason = "bypass set covers entire address space")
        }

        // Replace the FIRST AllowedIPs line with the complement,
        // delete any subsequent AllowedIPs lines (we collapse
        // multi-line entries into one for readability).
        val replacement = "AllowedIPs = " + complement.joinToString(", ") { it.toCidrString() }
        // Remove from highest idx down so indices stay valid.
        for (idx in allowedIPLineIdxs.drop(1).reversed()) {
            lines.removeAt(idx)
        }
        lines[allowedIPLineIdxs[0]] = replacement

        Log.i(TAG, "WG split tunnel applied: ${effectiveBypass.size} bypass " +
                "CIDRs, ${complement.size} resulting AllowedIPs entries")
        return InjectResult(lines.joinToString("\n"), applied = true)
    }

    /**
     * OpenVPN patch: prepend route directives so the bypass CIDRs
     * land on the local default gateway instead of the tunnel.
     * Append + pull-filter ignore for any server-pushed routes
     * that match (so the server can't undo our bypass).
     */
    private fun patchOpenVpn(
        content: String,
        bypass: List<CidrMath.Cidr>
    ): InjectResult {
        val sb = StringBuilder()
        // pull-filter must come before any other route directives
        // for OpenVPN's parser to apply it. Ignore any server
        // pushed route for an IP within our bypass list.
        sb.appendLine("# Privycs split tunnel: client-side bypass routes")
        for (c in bypass) {
            // pull-filter pattern matches the start of the route
            // directive. Server typically pushes "route a.b.c.d
            // m.m.m.m" - prefix-match catches it.
            if (c.isV4) {
                sb.appendLine("pull-filter ignore \"route ${c.toCidrString().substringBefore('/')}\"")
            } else {
                sb.appendLine("pull-filter ignore \"route-ipv6 ${c.toCidrString().substringBefore('/')}\"")
            }
        }
        for (c in bypass) {
            if (c.isV4) {
                val ip = c.toCidrString().substringBefore('/')
                val mask = prefixToIPv4Mask(c.prefix)
                sb.appendLine("route $ip $mask net_gateway")
            } else {
                sb.appendLine("route-ipv6 ${c.toCidrString()} net_gateway_ipv6")
            }
        }
        sb.append(content)
        Log.i(TAG, "OpenVPN split tunnel applied: ${bypass.size} bypass CIDRs")
        return InjectResult(sb.toString(), applied = true)
    }

    /** Convert "/24" prefix to "255.255.255.0" netmask form. */
    private fun prefixToIPv4Mask(prefix: Int): String {
        val mask = if (prefix == 0) 0L else (0xFFFFFFFFL shl (32 - prefix)) and 0xFFFFFFFFL
        return "${(mask shr 24) and 0xFF}.${(mask shr 16) and 0xFF}.${(mask shr 8) and 0xFF}.${mask and 0xFF}"
    }
}
